fix: folder tree ui

leopuleo opened the pull request:
## Changes
With this PR we improve the `FolderTree` component: 
- feat: add "loading" state
- feat: add "empty" state
- fix: keep folders open after the user clicks on another folder
- feat: improve drag and drop position feedback
- feat: improve drag and drop folder at the root level
- fix: sort folders by their name

Related to [WEB-1865](https://webiny.atlassian.net/browse/WEB-1865)

### Screenshots

Sorting
![sorting](https://user-images.githubusercontent.com/2866531/205601293-67d3d529-5b5e-45f0-9f54-b188817ef617.png)

Empty List
![empty](https://user-images.githubusercontent.com/2866531/205601329-dbe24302-71c0-4b87-9d09-3ecb1de6fa0f.png)

Loading List
![loading](https://user-images.githubusercontent.com/2866531/205601355-2b9ac083-9a98-491f-9349-ed02f14b6311.jpg)


## How Has This Been Tested?
<!--- Please describe how you tested your changes. -->
TBD

## Documentation
Documentation within the package

leopuleo committed:
fix: sort by folder name

leopuleo committed:
fix: add placeholder to preview dragging position

leopuleo committed:
fix: initial open list

leopuleo committed:
fix: initial open list

leopuleo committed:
fix: open list on title click

leopuleo committed:
fix: ui styling

leopuleo committed:
fix: sort alphanumerical strings

leopuleo committed:
feat: manage loading

leopuleo committed:
chore: loader styling

leopuleo committed:
feat: manage empty state

leopuleo committed:
chore: folder typography

leopuleo committed:
feat: improve dnd into root folder

leopuleo committed:
docs: improve docs

adrians5j commented on packages/app-folders/src/components/Tree/Placeholder/index.tsx:
```diff
@@ -0,0 +1,15 @@
+}
+
+export const Placeholder = ({ depth }: Props) => {
+    const left = depth * 24 + 8;
```
Maybe we leave a comment here, explaining why the formula?

leopuleo replied:
Done it, thank you! 🙏

adrians5j commented on packages/app-folders/src/components/Tree/index.tsx:
```diff
@@ -93,110 +30,42 @@ export const FolderTree: React.FC<Props> = ({
-                        }}
-                        initialOpen={initialOpenList}
+            <Title title={title} onClick={onTitleClick} isDragging={isDragging} />
+            {!folders ? (
```
I'd say, for better readability, we do this above the `return` line.

Having ternary and then nested ternary operators is a bit hard to read.

leopuleo replied:
Done it, thank you!

adrians5j commented on packages/app-folders/src/components/Tree/Node/index.tsx:
```diff
@@ -19,8 +20,29 @@ type Props = {
+    onUpdateFolder,
+    onDeleteFolder
+}: Props) => {
     const indent = depth * 24 + 8;
```
Mentioned the not-commented formula in a prev comment.

leopuleo replied:
👍

adrians5j commented on packages/app-folders/README.md:
```diff
@@ -32,32 +63,48 @@ import { FolderTree } from "@webiny/app-folders";
 
-As you might notice, there is not `listLinks` method available from `useLinks()` hook: this is because on first mount, `listLinks` is called internally, which will either issue a network request, or load links from cache.
+### FolderDialogDelete
+`FolderDialogUpdate` component shows a dialog to allow users to delete an existing folder.
```
```suggestion
`FolderDialogDelete` component shows a dialog to allow users to delete an existing folder.
```

leopuleo replied:
👍

adrians5j commented on packages/app-folders/README.md:
```diff
@@ -21,9 +21,40 @@ yarn add @webiny/app-folders
 ## Usage
```
Tip (for future GH docs): when writing GH docs, you can also use the format in this README:
https://github.com/webiny/webiny-js/tree/fix/folder-tree/packages/wcp
https://github.com/webiny/webiny-js/tree/fix/folder-tree/packages/api-wcp
https://github.com/webiny/webiny-js/tree/fix/folder-tree/packages/app-wcp

We [discussed this](https://www.notion.so/webiny/Proposal-Packages-Docs-Reference-f6d9ed56af9a4ae9921e86228bc2d202) a long time ago, and kinda said that we want to use it, but never actually started implementing it. :)

adrians5j commented on packages/app-folders/README.md:
```diff
@@ -21,9 +21,40 @@ yarn add @webiny/app-folders
 ## Usage
 Make sure `FoldersProvider` has been included within the admin app.
```
```suggestion
Make sure `FoldersProvider` has been included within the Admin app.
```

adrians5j commented on packages/app-folders/src/components/Tree/List/index.tsx:
```diff
@@ -0,0 +1,214 @@
+            setTreeData(createTreeData(folders, focusedFolderId));
+            setInitialOpenList(createInitialOpenList(folders, openFolderIds, focusedFolderId));
+        }
+    }, [Object.assign({}, folders), focusedFolderId]);
```
Hm, can you explain `Object.assign({}, folders)` here?

Also, if really needed, `[{ ...folders }, focusedFolderId]` might be a bit easier to read.

leopuleo replied:
Yeah, that's enough 😄

adrians5j commented on packages/app-folders/src/components/Tree/List/index.tsx:
```diff
@@ -0,0 +1,214 @@
+    });
+};
+
+const createInitialOpenList = (
```
Can we move these outside of this file, into sibling files or a sibling `utils.ts` file?

leopuleo replied:
Done it, thank you!

adrians5j reviewed (changes requested):
A couple of nitpicks, nothing spectacular.

When it comes to UI testing...

1. Noticed folders disappeared when I clicked on a folder above. Should this happen? Shouldn't the expanded folders still be visible? I'd say yes, at least that's what I'm used to seeing in other tools.

https://user-images.githubusercontent.com/5121148/205629350-fdf5a290-6d6d-4007-8bf0-c63df927e515.mov

2. Also, let's update the message here:
![image](https://user-images.githubusercontent.com/5121148/205620746-b693fec9-df88-4aaf-b78f-b375900a5388.png)

...to: 
![image](https://user-images.githubusercontent.com/5121148/205621670-93b28b5d-69e4-47f7-9b5c-62086c6742ff.png)

3. Can we have slug auto-generated here, just so the user doesn't have to type it all the time? 
![image](https://user-images.githubusercontent.com/5121148/205628123-321efdc1-b977-402e-a634-e91da6920294.png)

4. Numbers should be allowed in folder name, no? Am I missing something? 🤔 
![image](https://user-images.githubusercontent.com/5121148/205628488-3b609740-01d7-4895-ba04-854afba83994.png)

5. I think there's a bug here - folders disappear when clicking the edit button:

https://user-images.githubusercontent.com/5121148/205629232-b82fedac-2c2a-475f-a8b9-6eedd9b28280.mov

That's if for now I'd say. :) 
If you feel there's no time for some of the nitpicks, please then just write them down in Jira or somewhere.

leopuleo committed:
chore: merge next

leopuleo committed:
docs: add indent inline documentation

leopuleo committed:
chore: remove Object.assign

leopuleo committed:
feat: extract render list into fn

leopuleo committed:
docs: update docs

leopuleo committed:
docs: fix example

leopuleo committed:
feat: extract List methods into separated file

leopuleo committed:
chore: merge next

leopuleo commented:
Hi @adrians5j,
thanks for the review. About your comment here above: 

> 1. Noticed folders disappeared when I clicked on a folder above. Should this happen? Shouldn't the expanded folders still be visible? I'd say yes, at least that's what I'm used to seeing in other tools.

It should act as you describe, but I could not reproduce the bug 😢 

> 2. Also, let's update the message here

I agree with the changes, I added to this [subtask](https://webiny.atlassian.net/browse/WEB-1878).

> 3. Can we have slug auto-generated here, just so the user doesn't have to type it all the time?

It would be a nice feature, I added a [new subtask](https://webiny.atlassian.net/browse/WEB-1940) to address this.

> 4. Numbers should be allowed in folder name, no? Am I missing something? 🤔

Absolutely, this has been already fixed by this PR -> #2812 

> 5. I think there's a bug here - folders disappear when clicking the edit button:

Nice spot! I addressed the issue with this PR -> #2826

adrians5j commented on packages/app-folders/src/components/Tree/index.tsx:
```diff
@@ -93,110 +30,50 @@ export const FolderTree: React.FC<Props> = ({
+    const renderList = () => {
+        if (!folders) {
+            return <Loader />;
+        } else if (folders.length > 0) {
```
No need for `else if` here. We can do the following:

```ts
if (!folders) {
   return <Loader />;
} 

if (folders.length > 0) { ...
```

Makes the code a bit more readable. https://www.samanthaming.com/tidbits/23-no-else-return/

leopuleo replied:
Definitely more readable 😄

adrians5j commented on packages/app-folders/src/components/Tree/List/index.tsx:
```diff
@@ -0,0 +1,153 @@
+            setTreeData(createTreeData(folders, focusedFolderId));
+            setInitialOpenList(createInitialOpenList(folders, openFolderIds, focusedFolderId));
+        }
+    }, [{ ...folders }, focusedFolderId]);
```
Looks nicer, but I still have a question: why do we need to spread here? I don't see the reason to do it, since we're doing deep compare. Maybe I'm missing smt here? 🤔

leopuleo replied:
I'm adding a comment to explain it: this need more investigation in order to remove the object spreading

adrians5j reviewed (changes requested):
Two minor things. 

All other things looking good. 

Regarding issue 1 - could not repro that one either.

Thank you! :)

leopuleo committed:
fix: improve renderList condition handling

leopuleo committed:
fix: remove folders spread into useDeepCompareEffect deps

leopuleo committed:
docs: add a comment to explain the folders array spread

adrians5j reviewed (approved):
Thanks!
